plugin/pkg/shoot/quotavalidator/admission.go

```diff
@@ -83,15 +117,485 @@ func (h *RejectShootIfQuotaExceeded) Admit(a admission.Attributes) error {
+	// Pass if the shoot is intended to get deleted
+	if shoot.DeletionTimestamp != nil && shoot.Annotations[common.ConfirmationDeletionTimestamp] != "" {
+		return nil
+	}
+
+	var (
+		oldShoot         *garden.Shoot
+		maxShootLifetime *int
+		checkLifetime    = false
+		checkQuota       = false
+	)
+
+	if a.GetOperation() == admission.Create {
+		checkQuota = true
+	}
+
+	if a.GetOperation() == admission.Update {
+		oldShoot, ok = a.GetOldObject().(*garden.Shoot)
+		if !ok {
+			return apierrors.NewBadRequest("could not convert resource into Shoot object")
+		}
+		cloudProvider, err := helper.DetermineCloudProviderInShoot(shoot.Spec.Cloud)
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+
+		checkQuota = quotaVerificationNeeded(*shoot, *oldShoot, cloudProvider)
+		checkLifetime = lifetimeVerificationNeeded(*shoot, *oldShoot)
+	}
+
+	secretBinding, err := h.secretBindingLister.SecretBindings(shoot.Namespace).Get(shoot.Spec.Cloud.SecretBindingRef.Name)
+	if err != nil {
+		return apierrors.NewInternalError(err)
+	}
+
+	// Quotas are cumulative, means each quota must be not exceeded that the admission pass.
+	for _, quotaRef := range secretBinding.Quotas {
+		quota, err := h.quotaLister.Quotas(quotaRef.Namespace).Get(quotaRef.Name)
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+
+		// Get the max clusterLifeTime
+		if checkLifetime && quota.Spec.ClusterLifetimeDays != nil {
+			if maxShootLifetime == nil {
+				maxShootLifetime = quota.Spec.ClusterLifetimeDays
+			}
+			if *maxShootLifetime > *quota.Spec.ClusterLifetimeDays {
+				maxShootLifetime = quota.Spec.ClusterLifetimeDays
+			}
+		}
+
+		if checkQuota {
+			exceededMetrics, err := h.isQuotaExceeded(*shoot, *quota)
+			if err != nil {
+				return apierrors.NewInternalError(err)
+			}
+			if exceededMetrics != nil {
+				message := ""
+				for _, metric := range *exceededMetrics {
+					message = message + metric.String() + " "
+				}
+				return admission.NewForbidden(a, fmt.Errorf("Quota limits exceeded. Unable to allocate further %s", message))
+			}
+		}
+	}
+
+	// Admit Shoot lifetime changes
+	if lifetime, exists := shoot.Annotations[common.ShootExpirationTimestamp]; checkLifetime && exists && maxShootLifetime != nil {
+		var (
+			plannedExpirationTime   time.Time
+			oldExpirationTime       time.Time
+			calulatedExpirationTime time.Time
+		)
+
+		plannedExpirationTime, err = time.Parse(time.RFC3339, lifetime)
+		if err != nil {
+			return apierrors.NewInternalError(err)
+		}
+
+		if lifetime, exists := oldShoot.Annotations[common.ShootExpirationTimestamp]; exists {
+			oldExpirationTime, err = time.Parse(time.RFC3339, lifetime)
+			if err != nil {
+				return apierrors.NewInternalError(err)
+			}
+		} else {
+			oldExpirationTime = shoot.CreationTimestamp.Time
+		}
+
+		calulatedExpirationTime = oldExpirationTime.Add(time.Duration(*maxShootLifetime*24) * time.Hour)
```

Comment:
See above comment.

Reply:
Fixed accordingly. Here we have the special case that the old version of the Shoot does not have a clusterDeletion timestamp annotated yet. Therefore we have to calculate the oldLifetime in this case.